### PR TITLE
Add group by function overload to ġ

### LINKED ISSUE #964: Group by builtin
(can be a monograph)

A builtin that takes a function and a list, and groups the values of a list by the result when the function is applied to each of them.

### PR DESCRIPTION
**Pull request by Lyxal**
Closes #964

### CODE REVIEW

File: vyxal/helpers.py [hunk 1]
```diff
@@ -355,6 +355,25 @@ def graph_distance(
         )
 
 
+def group_by_function(
+    lst: VyList, function: types.FunctionType, ctx: Context
+) -> LazyList:
+    """Group a list of elements by a function"""
+
+    @lazylist
+    def gen():
```
Comment (cgccuser) [inline vyxal/helpers.py:364]: There’s no need to make this a LazyList. It goes over the entire list before yielding all the values at once. If you like, I could try an infinite-list-friendly implementation later but it’s probably not going to work

**Review by cgccuser — CHANGES_REQUESTED** [review]

File: vyxal/helpers.py (hunk 1 above)
Comment (Steffan153) [inline vyxal/helpers.py:364]:
```suggestion
    """Group a list of elements by a function"""

    def gen():
```

**Review by Steffan153 — APPROVED** [review]

File: vyxal/helpers.py [hunk 2]
```diff
@@ -355,6 +355,24 @@ def graph_distance(
...
+        ret = {}
+        for el in lst:
+            key = safe_apply(function, el, ctx=ctx)
+            if key in ret:
+                ret[key].append(el)
+            else:
+                ret[key] = [el]
+        yield from ret.values()
+
+    return gen()
```
Comment (cgccuser) [inline vyxal/helpers.py:373]: I meant to take out the `gen` function entirely, you sussy bakas
```suggestion
    ret = {}
    for el in lst:
        key = safe_apply(function, el, ctx=ctx)
        if key in ret:
            ret[key].append(el)
        else:
            ret[key] = [el]
    return ret.values()
```

**Review by cgccuser — CHANGES_REQUESTED** [review]

File: vyxal/helpers.py [hunk 3]
```diff
@@ -355,6 +355,21 @@ def graph_distance(
...
+    """Group a list of elements by a function"""
+
+    ret = {}
+    for el in lst:
+        key = safe_apply(function, el, ctx=ctx)
+        if key in ret:
+            ret[key].append(el)
+        else:
+            ret[key] = [el]
+    return ret.values()
```
Comment (cgccuser) [inline vyxal/helpers.py:370]:
```suggestion
    return list(ret.values())
```

**Review by cgccuser — APPROVED** [review]

**Pull request merged**